client-runtime/src/main/java/com/microsoft/rest/v2/http/NettyClient.java

```diff
@@ -353,98 +360,206 @@ public void onComplete() {
-                        }
+        void chunkCompleted() {
+            if (done) {
+                RxJavaPlugins.onError(new IllegalStateException("chunk completion event occurred after done is true"));
+                return;
+            }
+            chunkRequested.set(false);
+            drain();
+        }
 
-                        if (chunksRequested > 0 && !isCanceled) {
-                            handlerSubscription.request(chunksRequested);
-                        }
-                    } finally {
-                        draining = false;
-                    }
-                }
+        void onError(Throwable cause) {
+            if (done) {
+                RxJavaPlugins.onError(cause);
+            }
+            done = true;
+            err = cause;
+            drain();
+        }
+        
+        void channelInactive() {
+            if (!done) {
+                done = true;
+                err = new IOException("channel inactive");
+                drain();
             }
         }
 
-        @Override
-        public void cancel() {
-            isCanceled = true;
-            handlerSubscription.cancel();
+        //
+        // PRIVATE METHODS
+        //
 
-            currentEventLoop.execute(() -> {
-                while (!queuedContent.isEmpty()) {
-                    queuedContent.remove().release();
+        private void requestChunkOfByteBufsFromUpstream() {
+            channelSubscription.request(1);
+        }
+        
+        private void drain() {
+            if (wip.getAndIncrement() == 0) {
+                // need to check cancelled even if there are no requests
+                if (cancelled) {
+                    releaseQueue();
+                    return;
                 }
-            });
+                int missed = 1;
+                long r = requested.get();
+                while (true) {
+                    long e = 0;
+                    while (e != r) {
+                        // Note that an error can shortcut the emission of content that is currently on the queue.
+                        // This is probably desirable generally because it prevents work that being done downstream 
+                        // that might be thrown away anyway due to the error. 
+                        // TODO discuss with project members
+                        Throwable error = err;
+                        if (error != null) {
+                            releaseQueue();
+                            channelSubscription.cancel();
+                            subscriber.onError(error);
+                            return;
+                        }
+                        HttpContent o = queue.poll();
+                        if (o != null) {
+                            e++;
+                            if (emitContent(o)) {
+                                return;
+                            }
+                        } else {
```

Comment:
It looks like the drain loop continues even when the queue is empty (`poll()` returns null). Is that the right thing to do?

This seems to be causing a hanging issue where `responseEmitter.onSuccess()` synchronously subscribes to the `ResponseContentFlowable` and requests the max amount. Then, this drain loop runs forever because the Netty thread never frees up to service the `channelRead` and put content into the queue.

Reply:
Spot on. There needs to be a `break` there.

Reply:
Excellent-- my tests look all good locally with this latest change.